### PR TITLE
apple, server, core: apple billing

### LINKED ISSUE #984: apple pay server support

### LINKED ISSUE #985: use apple pay for payments

### PR DESCRIPTION
**Pull request by SmailBarkouch**
Apple users can now purchase premium via App Store billing. This PR is split into two parts:
- Front-end UI work on the apple clients. The majority of this UI is found in the `clients/apple/Shared` folder.
- Back-end UI work on core and server. This includes validations when first upgrading to premium, and processing app store server notifications.

### New Server Endpoints
- `/upgrade-account-app-store`: Accepts the original transaction id (first transaction id), app account token (used to link a lockbook account with an App Store account), and an encoded receipt. The encoded receipt is sent to apple to for validation, and the returned details will be checked up against the app account token and original transaction id.
- `/app_store_notification_webhook`: Processes app store server notifications.

### Code Organization
- `server/.../billing/app_store_client.rs`: Builds the requests that are made to apple's servers. This includes creating a JWS for authentication.
- `server/.../billing/app_store_service.rs`: Helper functions for `upgrade_account_app_store` and `app_store_notification_webhook`.
- `server/.../billing/app_store_model.rs`: The data types that are needed for parsing the responses and request bodys' of apple requests.

### Progress
- [x] Create iOS UI
- [x] Create MacOS UI
- [x] Implement `upgrade_account_app_store` endpoint
- [x] Implement `app_store_notifications_webhook` endpoint
- [x] Test billing for iOS
- [x] Test billing for MacOS
- [x] Test edge case billing operations on app store server notifications
- [x] screenshots in docs

fixes #984
fixes #985

### CODE REVIEW

File: clients/apple/Shared/Stateful Logic/BillingService.swift [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    
+    init(_ core: LockbookApi) {
+        self.core = core
+        
+        if case .success(_) = core.getAccount() {
+            Task {
+                await launchBillingBackgroundTasks()
+            }
+        }
+    }
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/BillingService.swift:44]: What I'm thinking here is: I don't want the app to do this every time I open it. Ideally it only does this when I'm doing the billing flow. Costs some data, some battery, and maybe adds a hair of latency to startup (i know it's async). Can't we fire this off at some later point in the flow than every app startup?

File: clients/egui/src/widgets/subscription.rs [hunk 2]
```diff
@@ -32,6 +32,7 @@ fn subscription_info(
         Some(info) => match &info.payment_platform {
             Stripe { card_last_4_digits } => draw_stripe(ui, card_last_4_digits),
             GooglePlay { account_state } => draw_google_play(ui, account_state),
+            AppStore { .. } => todo!(),
```
Comment (Parth) [inline clients/egui/src/widgets/subscription.rs:35]: I think not showing this for now is fine, but `panic`'ing here is prob not the move, let's create a tracking issue for this and not panic.

File: server/server/Cargo.toml [hunk 3]
```diff
@@ -38,6 +38,9 @@ tracing = "0.1.5"
 tracing-subscriber = "0.3.9"
 tracing-appender = "0.2"
 itertools = "0.10.1"
+reqwest = { version = "0.11", features = ["json"] }
+jsonwebtoken = "*"
```
Comment (Parth) [inline server/server/Cargo.toml:42]: why *?

File: server/server/src/billing/app_store_client.rs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                            |err| {
+                                internal!(
+                                    "Cannot decode apple jwt payload: {:?}, err: {:?}",
+                                    part,
+                                    err
+                                )
+                            },
+                        )?)?;
+                    return Ok((last_trans.clone(), trans_info));
+                }
```
Comment (Parth) [inline server/server/src/billing/app_store_client.rs:106]: safe to ignore the rest of this data in the loop?

File: server/server/src/billing/app_store_client.rs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            }
+
+            Err(internal!("No usable data returned from apple subscriptions statuses endpoint despite assumed match. resp_body: {:?}, monthly_sub_group: {}", sub_status, config.monthly_sub_group_id))
+        }
+        400 | 404 => {
+            let resp_body = resp.text().await.ok();
+            info!(
```
Comment (Parth) [inline server/server/src/billing/app_store_client.rs:113]: perhaps this is `error!` worthy?

File: server/server/src/billing/billing_service.rs [hunk 6]
```diff
@@ -73,6 +79,67 @@ fn release_subscription_profile<T: Debug>(
...
+    debug!("Upgrading the account of a user through app store billing");
+
+    if let Some(owner) = server_state
+        .index_db
+        .app_store_ids
+        .get(&request.app_account_token)?
+    {
+        if let Some(other_account) = server_state.index_db.accounts.get(&owner)? {
+            if other_account.billing_info.billing_platform.is_some() {
+                return Err(ClientError(UpgradeAccountAppStoreError::AppStoreAccountAlreadyLinked));
```
Comment (Parth) [inline server/server/src/billing/billing_service.rs:102]: wouldn't we return this error if they had `Strip`? Prob need to rename the error

File: server/server/src/billing/billing_service.rs [hunk 7]
```diff
@@ -73,6 +79,67 @@ fn release_subscription_profile<T: Debug>(
...
+            }
+        }
+    }
+
+    let expires = verify_details(
```
Comment (Parth) [inline server/server/src/billing/billing_service.rs:107]: I would refer to this as `app_store_service::verify_details` but nbd

Comment (Parth) [inline server/server/src/billing/billing_service.rs:107]: Do we need to check this `expires` for validity?

File: server/server/src/schema.rs [hunk 8]
```diff
@@ -45,7 +45,8 @@ pub mod v2 {
             file_children: <Uuid, HashSet<Uuid>>,
             sizes: <Uuid, u64>,
             google_play_ids: <String, Owner>,
-            stripe_ids: <String, Owner>
+            stripe_ids: <String, Owner>,
+            app_store_ids: <String, Owner>
```
Comment (Parth) [inline server/server/src/schema.rs:49]: This needs to end up in a v3 with a migration, see the rest of this file. Also see how this is gonna interact with the proposed schema changes [here](https://github.com/lockbook/lockbook/issues/1296#issuecomment-1368230780)

cc: @ad-tra 

I'm down to not do it in this PR and let @ad-tra just do it in his PR

File: server/server/src/lib.rs [hunk 9]
```diff
@@ -1,3 +1,5 @@
+extern crate core;
```
Comment (Parth) [inline server/server/src/lib.rs:1]: prob an accidental auto import

**Review by Parth — CHANGES_REQUESTED** [review]

File: server/server/src/billing/app_store_client.rs (hunk 4 above)
Comment (SmailBarkouch) [reply server/server/src/billing/app_store_client.rs:107]: Yes it is, we are only looking for a particular subscription. There should not be other things in this loop regardless though.

File: server/server/src/billing/billing_service.rs (hunk 6 above)
Comment (SmailBarkouch) [reply server/server/src/billing/billing_service.rs:102]: I ended up `if let`ing an app store subscription, so there is no need to rename. Also, it is what should be done since we only care if it is an app store subscription.

File: server/server/src/billing/billing_service.rs (hunk 7 above)
Comment (SmailBarkouch) [reply server/server/src/billing/billing_service.rs]: We do not since in `verify_details`, the subscription status endpoint we make a request to tells us if the subscription has expired or not.

File: server/server/src/billing/billing_service.rs (hunk 6 above)
Comment (Parth) [reply server/server/src/billing/billing_service.rs:102]: What if they have another platform like stripe?

Comment (SmailBarkouch) [reply server/server/src/billing/billing_service.rs:102]: It doesn't matter if they have another platform so long as the platform is not apple. But on second thought, it is possible that the platform is apple but not the same apple account. I will add a check for the origin transaction id.

If the account happened to be stripe it would not matter because all this check is looking for is if an apple account is not being used by someone else. An apple account can only pay for one subscription.

**Review by Parth — APPROVED** [review]
Congrats dawg
